### PR TITLE
Unused Beans and Why We Remove Them

### PR DESCRIPTION
**Pull request by mkouba**
This is a follow-up of https://github.com/quarkusio/quarkus/pull/18970 where a more strict algorithm was introduced and docs were updated: https://quarkus.io/version/main/guides/cdi-reference#remove_unused_beans

### CODE REVIEW

File: _posts/2021-09-23-unused-beans.adoc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+title: 'Unused Beans and Why We Remove Them'
+date: 2021-09-23
+tags: cdi arc 
+synopsis: Why does Quarkus detect and remove unused beans by default?
+author: mkouba
+---
+
+Quarkus is a build-time oriented stack, i.e. it tries to do as much as possible at build time to improve the startup time and memory usage of your application.
+However, it's not always so easy and straightforward.
+Many existing frameworks and libraries do not take this design choice into account.
```
Comment (manovotn) [inline _posts/2021-09-23-unused-beans.adoc:12]: The last three sentenced are kind of clunky to read. I'd merge two of them together, so something like:
> However, it's not always so easy and straightforward as many existing frameworks and libraries do not take this design choice into account. And CDI is one of them.

File: _posts/2021-09-23-unused-beans.adoc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+And CDI is one of them.
+
+== The Why
```
Comment (manovotn) [inline _posts/2021-09-23-unused-beans.adoc:15]: This kind of title is IMO nice if your other titles are in a similar fashion (e.g. The What, The Why, The How....) which isn't the case here.
You could use something like `Motivation` or `Why Bother`.

File: _posts/2021-09-23-unused-beans.adoc [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+Speaking of generated classes.
+The bean discovery, validation and wiring all components together - all this is perfomed at build time.
+Quarkus then stores the collected metadata in the bytecode, i.e. for each bean one or more classes are generated.
+In order to fullfill some basic CDI API requirements, a bean has at least a corresponding `javax.enterprise.inject.spi.Bean` implementation.
+If it's a normal scoped bean then a client proxy class must be also generated.
+Finally, intercepted and decorated beans require some more internal constructs.
+
+Imagine that your application contains 50 beans that are actually not used anywhere.
+If they're normal scoped (e.g. has `@ApplicationScoped`) and intercepted (e.g. annotated with `@Transactional`) you can expect more than 150 generated classes.
```
Comment (manovotn) [inline _posts/2021-09-23-unused-beans.adoc:32]: > e.g. has `@ApplicationScoped`

Maybe instead say `any other scope than @Dependent` to underline how many beans are affected by it?

File: _posts/2021-09-23-unused-beans.adoc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+And these classes are completely useless.
+Still, the container would have to instantiate and hold a reference on those 50+ bean metadata classes.
+Needless to say, that the bean classes and any referenced classes cannot be a subject of dead code elimination when a native image is generated.
```
Comment (manovotn) [inline _posts/2021-09-23-unused-beans.adoc:35]: > cannot be a subject of dead 

I think the correct term is `subject to dead code elimination`?

File: _posts/2021-09-23-unused-beans.adoc [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+== What’s Actually Removed?
+
+Quarkus first identifies so-called _unremovable_ beans that form the roots in the dependency tree.
+These classes very often represent the application _entry points_.
+A good example is a JAX-RS resource class.
+Another example is a bean which declares a `@Scheduled` method.
+
+An _unremovable_ bean:
+- is excluded from removal by an extension, or
```
Comment (manovotn) [inline _posts/2021-09-23-unused-beans.adoc:46]: It might be worth mentioning `@Unremovable` here? Most users won't write extensions so talking about the annotation might make more sense.

File: _posts/2021-09-23-unused-beans.adoc [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+- is not unremovable, and
+- is not eligible for injection to any injection point in the dependency tree, and
+- does not declare any producer which is eligible for injection to any injection point in the dependency tree, and
+- is not eligible for injection into any `javax.enterprise.inject.Instance` or `javax.inject.Provider` injection point.
+
+Finally, unused interceptors and decorators are not associated with any bean.
+
+TIP: When using the dev mode (e.g. running `./mvnw quarkus:dev`), you can see more information about which beans are being removed in the Dev UI.
+
+== The Main Drawback
```
Comment (manovotn) [inline _posts/2021-09-23-unused-beans.adoc:60]: I'd remove definite article from the title

File: _posts/2021-09-23-unused-beans.adoc (hunk 5 above)
Comment (Ladicek) [inline _posts/2021-09-23-unused-beans.adoc:46]: I assume that "an extension" here means Quarkus extension, and so this case covers the examples from previous paragraph. If so, I think this connection is far from obvious. I'd rephrase to something like this:

> Quarkus first identifies so-called _unremovable_ beans that form the roots in the dependency tree.
> Unremovable beans are beans that:
> - declare an observer method, or
> - have a name designated via `@Named`, or
> - are otherwise identified by Quarkus extensions.
>
> The last point is probably most important, because this is how application _entry points_ are made unremovable.
> A good example is a JAX-RS resource class, identified by the RESTEasy extension.
> Another example is a bean which declares a `@Scheduled` method, identified by the Scheduler extension.

File: _posts/2021-09-23-unused-beans.adoc (hunk 2 above)
Comment (mkouba) [reply _posts/2021-09-23-unused-beans.adoc:15]: I like "Why Bother".

File: _posts/2021-09-23-unused-beans.adoc (hunk 4 above)
Comment (mkouba) [reply _posts/2021-09-23-unused-beans.adoc:35]: You're right!

File: _posts/2021-09-23-unused-beans.adoc (hunk 5 above)
Comment (mkouba) [reply _posts/2021-09-23-unused-beans.adoc:46]: Hm, yes we should rephrase this sentence...

File: _posts/2021-09-23-unused-beans.adoc (hunk 3 above)
Comment (mkouba) [reply _posts/2021-09-23-unused-beans.adoc:32]: Actually, we would have to use `any other scope than @Dependent and @Singleton` which is a bit long... I'll keep the `@ApplicationScoped` which is very common and understandable...

**Comment by mkouba** [comment]
@Ladicek @manovotn Thanks a lot for your feedback, much appreciated!

File: _posts/2021-09-23-unused-beans.adoc [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+- declare an observer method, or
+- have a name designated via `@Named`, or
+- are annotated with `@io.quarkus.arc.Unremovable`, or
+- are excluded via the `quarkus.arc.unremovable-types` config property, or
+- are otherwise identified by Quarkus extensions.
+
+The last point is probably most important, because this is how application entry points are made unremovable.
+A good example is a JAX-RS resource class, identified by the RESTEasy extension.
+Another example is a bean which declares a @Scheduled method, identified by the Scheduler extension.
```
Comment (manovotn) [inline _posts/2021-09-23-unused-beans.adoc:50]:
```suggestion
Another example is a bean which declares a `@Scheduled` method, identified by the Scheduler extension.
```

**Review by manovotn — APPROVED** [review]
Added one tiny comment, otherwise LGTM!

**Comment by mkouba** [comment]
CC @maxandersen @geoand. I have no idea who should be the one to push the green button ;-)

**Comment by geoand** [comment]
It doesn't have to be Max or I :).

If the blog post has been peer reviewed, feel free to merge it.

**Review by Ladicek — APPROVED** [review]

**Comment by Ladicek** [comment]
I think it's good! :-)
